x/bsonx/bsoncore/document.go

```diff
@@ -264,58 +261,73 @@ func (d Document) DebugString() string {
 // String outputs an ExtendedJSON version of Document. If the document is not valid, this method
 // returns an empty string.
 func (d Document) String() string {
-	return d.StringN(math.MaxInt)
+	str, _ := d.stringN(0)
+	return str
 }
 
 // StringN stringifies a document upto N bytes
-func (d Document) StringN(n int) string {
-	if len(d) < 5 || n <= 0 {
-		return ""
+func (d Document) StringN(n int) (string, bool) {
+	if n <= 0 {
+		if l, _, ok := ReadLength(d); !ok || l < 5 {
+			return "", false
+		}
+		return "", true
 	}
+	return d.stringN(n)
+}
 
-	var buf strings.Builder
+// stringN stringify a document. If N is larger than 0, it will truncate the string to N bytes.
+func (d Document) stringN(n int) (string, bool) {
+	length, rem, ok := ReadLength(d)
+	if !ok || length < 5 {
+		return "", false
+	}
+	length -= (4 /* length bytes */ + 1 /* final null byte */)
 
+	var buf strings.Builder
 	buf.WriteByte('{')
 
-	length, rem, _ := ReadLength(d)
-	length -= 4
-
+	var truncated bool
 	var elem Element
-	var ok bool
-
+	var str string
 	first := true
-	truncated := false
-
-	if n > 0 {
-		for length > 1 {
-			if !first {
-				buf.WriteByte(',')
-			}
-			elem, rem, ok = ReadElement(rem)
-			length -= int32(len(elem))
-			if !ok {
-				return ""
-			}
-
-			str := elem.StringN(n)
-			if buf.Len()+len(str) > n {
-				truncatedStr := bsoncoreutil.Truncate(str, n-buf.Len())
-				buf.WriteString(truncatedStr)
-
+	for length > 0 && !truncated {
```

Comment:
[nit] The loop logic is hard to follow, suggest adding comments. I think the following are sufficient:

1. determine remaining budget `l := n - buf.Len()`
2. If `buf.Len() >= n`, then mark truncated and break 
3. if `!first`, then write comma, then decrement `l` and possibly break 
4. ReadElement, subtract its full length, exit on `!ok`
5. Delegate to `elem.stringN(l)`, write its str, record its truncated flag